farasa/__base.py

```diff
@@ -115,20 +115,19 @@ def _check_java_version(self):
             print(proc_err)
             raise Exception("We could not check for java version on the machine. Please make sure you have installed Java 1.7+ and add it to your PATH.")
 
-
     def _run_task(self, btext):
         assert btext is not None
-        with tempfile.NamedTemporaryFile(dir=f'{self.__base_dir}/tmp',) as itmp,\
-                    tempfile.NamedTemporaryFile(dir=f'{self.__base_dir}/tmp',) as otmp:
+        with tempfile.NamedTemporaryFile(dir=f'{self.__base_dir}/tmp', delete=False) as itmp, \
+                tempfile.NamedTemporaryFile(dir=f'{self.__base_dir}/tmp', delete=False) as otmp:
```

Comment:
Thanks for raising the attention to this bug. Also, thanks for passing by the documentation to dig more about the problem. I found out setting `delete= False`, albeit solving the issue on Windows, it will not delete the file once it is closed. Hence, if we are going to adhere to this solution, we need to manually delete these temp files ourselves. Otherwise, the temp directory will blow up with these temp files. What do you think? If you have any other ideas please advise.

Reply:
You are right, i also found out there is temp file left in tmp directory when the program is terminated, i will do more study about this, and i will find another Ubuntu server to do test, then we will got 3 platform (macOS/Windows/Linux) test result, for good compatibility

Reply:
About this issue, i study some hours to find solution, i think there are 2 options:

1. use `try...finally`, remove these temp file after done each task
2. I found out `django` has same problem, and they write an external class to solve this issue [https://github.com/django/django/blob/master/django/core/files/temp.py](https://github.com/django/django/blob/master/django/core/files/temp.py)

solution 1

```python
    def _run_task(self, btext):
        assert btext is not None
        tmpdir = str(self.__base_dir / 'tmp')
        itmp = tempfile.NamedTemporaryFile(dir=tmpdir, delete=False)
        otmp = tempfile.NamedTemporaryFile(dir=tmpdir, delete=False)
        try:
            itmp.write(btext)
            # https://stackoverflow.com/questions/46004774/python-namedtemporaryfile-appears-empty-even-after-data-is-written
            itmp.flush()
            proc = subprocess.run(self.__APIs[self.task] + ['-i', itmp.name, '-o', otmp.name], capture_output=True)
            if proc.returncode == 0:
                result = otmp.read().decode('utf8').strip()
            else:
                print("error occurred! stdout: ", proc.stdout, ' stderr: ', proc.stderr)
                print("return code:", proc.returncode)
                raise Exception('Internal Error occurred!')
        finally:
            itmp.close()
            otmp.close()
            os.unlink(itmp.name)
            os.unlink(otmp.name)
        return result
```

solution 2

1. add python file `util/temp_file.py` (from project django)

```python
"""
The temp module provides a NamedTemporaryFile that can be reopened in the same
process on any platform. Most platforms use the standard Python
tempfile.NamedTemporaryFile class, but Windows users are given a custom class.
This is needed because the Python implementation of NamedTemporaryFile uses the
O_TEMPORARY flag under Windows, which prevents the file from being reopened
if the same flag is not provided [1][2]. Note that this does not address the
more general issue of opening a file for writing and reading in multiple
processes in a manner that works across platforms.
The custom version of NamedTemporaryFile doesn't support the same keyword
arguments available in tempfile.NamedTemporaryFile.
1: https://mail.python.org/pipermail/python-list/2005-December/336957.html
2: https://bugs.python.org/issue14243
"""

import os
import tempfile

__all__ = ('NamedTemporaryFile', 'gettempdir',)


class FileProxyMixin:
    """
    A mixin class used to forward file methods to an underlaying file
    object.  The internal file object has to be called "file"::
        class FileProxy(FileProxyMixin):
            def __init__(self, file):
                self.file = file
    """

    encoding = property(lambda self: self.file.encoding)
    fileno = property(lambda self: self.file.fileno)
    flush = property(lambda self: self.file.flush)
    isatty = property(lambda self: self.file.isatty)
    newlines = property(lambda self: self.file.newlines)
    read = property(lambda self: self.file.read)
    readinto = property(lambda self: self.file.readinto)
    readline = property(lambda self: self.file.readline)
    readlines = property(lambda self: self.file.readlines)
    seek = property(lambda self: self.file.seek)
    tell = property(lambda self: self.file.tell)
    truncate = property(lambda self: self.file.truncate)
    write = property(lambda self: self.file.write)
    writelines = property(lambda self: self.file.writelines)

    @property
    def closed(self):
        return not self.file or self.file.closed

    def readable(self):
        if self.closed:
            return False
        if hasattr(self.file, 'readable'):
            return self.file.readable()
        return True

    def writable(self):
        if self.closed:
            return False
        if hasattr(self.file, 'writable'):
            return self.file.writable()
        return 'w' in getattr(self.file, 'mode', '')

    def seekable(self):
        if self.closed:
            return False
        if hasattr(self.file, 'seekable'):
            return self.file.seekable()
        return True

    def __iter__(self):
        return iter(self.file)


if os.name == 'nt':
    class TemporaryFile(FileProxyMixin):
        """
        Temporary file object constructor that supports reopening of the
        temporary file in Windows.
        Unlike tempfile.NamedTemporaryFile from the standard library,
        __init__() doesn't support the 'delete', 'buffering', 'encoding', or
        'newline' keyword arguments.
        """

        def __init__(self, mode='w+b', bufsize=-1, suffix='', prefix='', dir=None):
            fd, name = tempfile.mkstemp(suffix=suffix, prefix=prefix, dir=dir)
            self.name = name
            self.file = os.fdopen(fd, mode, bufsize)
            self.close_called = False

        # Because close can be called during shutdown
        # we need to cache os.unlink and access it
        # as self.unlink only
        unlink = os.unlink

        def close(self):
            if not self.close_called:
                self.close_called = True
                try:
                    self.file.close()
                except OSError:
                    pass
                try:
                    self.unlink(self.name)
                except OSError:
                    pass

        def __del__(self):
            self.close()

        def __enter__(self):
            self.file.__enter__()
            return self

        def __exit__(self, exc, value, tb):
            self.file.__exit__(exc, value, tb)


    NamedTemporaryFile = TemporaryFile
else:
    NamedTemporaryFile = tempfile.NamedTemporaryFile

gettempdir = tempfile.gettempdir

```

use this module

```python
from farasa.util.temp_file import NamedTemporaryFile

# ... other code

    def _run_task(self, btext):
        assert btext is not None
        with NamedTemporaryFile(dir=str(self.__base_dir / 'tmp')) as itmp, \
                NamedTemporaryFile(dir=str(self.__base_dir / 'tmp')) as otmp:
            itmp.write(btext)
            itmp.flush()  # https://stackoverflow.com/questions/46004774/python-namedtemporaryfile-appears-empty-even-after-data-is-written
            proc = subprocess.run(self.__APIs[self.task] + ['-i', itmp.name, '-o', otmp.name], capture_output=True)
            if proc.returncode == 0:
                result = otmp.read().decode('utf8').strip()
                return result
            else:
                print("error occurred! stdout: ", proc.stdout, ' stderr: ', proc.stderr)
                print("return code:", proc.returncode)
                raise Exception('Internal Error occured!')
```

known issue

via my test, if i terminating the task running in mode `standalone` manually (ctrl+c in CLI, or terminate button in PyCharm), it will still left temp file in temp directory

Reply:
Thanks for the effort you are putting @hefengxiang, I really appreciate it. I have a question here. Based on the research you did, does django class handles the issue of the other approach? If so, why not to adapt it?